### PR TITLE
Comp fixes

### PR DESCRIPTION
**Pull request by MarkusBagge**
### Make SignalPID2 obey set output limits.
The SignalPID2 have output limits as parameters. However these were not obeyed. Now they are.

### Add base unit Ohm. Changed ohm to Ohm in some models/components.
There is no quantity for resistance. Added here. Components/models that used "ohm" and "Ohm" inconsistently are changed to "Ohm".

### Make quantity AngularVelocity an alias of Frequency.
Obosolete. See "Use both AngularFrequency rad/s and Frequency Hz" and  "Resolve quantities from base units when adding ports and nodes. " below.

### New QuantityRegister::lookupQuantityByAlias
Hopsan GUI sometimes fails to interpret quantities, for instance when plotting. This is becuase the GUI does not handle quantity aliases. This fix translates all aliases to original quantites in port/nodes and will allow users to correctly select units in GUI.  

### Use both AngularFrequency rad/s and Frequency Hz
Revert most changes in "Make quantity AngularVelocity an alias of Frequency" above. (Sorry for the mess, didn't know how to do it...)

### Resolve quantities from base units when adding ports and nodes. 
This fix adds a method to resolve quantities from base units at the creation of port/nodes so that the GUI doesn't need to make guesses later on.

### CODE REVIEW

**Comment by peterNordin** [comment]
I know that my old code is "less then perfect" so I appreciate your help. 
I will need a few days to  review and test this, I is important to get it right since it changes core functionality, but I will write comments as I go maybe @robbr48 can participate in the discussion as well.

File: componentLibraries/defaultLibrary/Electric/ElectricACmachine.hpp [hunk 1]
```diff
@@ -167,7 +167,7 @@ class ElectricACmachine : public ComponentQ
         mpPel2=addPowerPort("Pel2","NodeElectric");
         mpPmr1=addPowerPort("Pmr1","NodeMechanicRotational");
         //Add inputVariables to the component
-            addInputVariable("freq","AC frequency","1/s",50,&mpfreq);
+            addInputVariable("freq","AC frequency","Hz",50,&mpfreq);
```
Comment (peterNordin) [inline componentLibraries/defaultLibrary/Electric/ElectricACmachine.hpp]: Could write Frequency instead of unit, also should be fixed in source modelica file by @petkr14  (also in other files)
When Quantity is known, it should be used instead of "base SI unit" (Quantities are fairly "new") so lots of old code needs to be updated.

File: HopsanCore/src/Quantities.cpp [hunk 2]
```diff
@@ -56,13 +55,18 @@ hopsan::QuantityRegister::QuantityRegister()
...
     registerQuantity("Density", "kg/m^3");
 
-    registerQuantity("Frequency", "rad/s");
+    registerQuantity("Frequency", "Hz");
+    registerQuantity("AngularVelocity", "rad/s");
     registerQuantity("Time", "s");
 
     registerQuantity("Temperature", "K");
 
+    registerQuantity("Resistance", "Ohm");
```
Comment (peterNordin) [inline HopsanCore/src/Quantities.cpp]: Why Ohm with a capital O? I don know myself, but some quick googling and Wikipediaing shows me that it seem to be written with lower-case ohm

File: HopsanCore/src/Quantities.cpp [hunk 3]
```diff
@@ -56,13 +55,18 @@ hopsan::QuantityRegister::QuantityRegister()
 
     registerQuantity("Density", "kg/m^3");
 
-    registerQuantity("Frequency", "rad/s");
+    registerQuantity("Frequency", "Hz");
```
Comment (peterNordin) [inline HopsanCore/src/Quantities.cpp]: Changing the base unit for a quantity may be dangerous, We need to verify what will happen to old models that have saved values in Frequency parameters. Will loading them still show and simulate the correct value.
For instance if it is loaded in HopsanGUI it may incorrectly re-scale the value to the new base unit. 

- [ ] figure out what may happen when old models are loaded when a quantity has changed base unit. It my be possible to auto update models as we have done in the passt. (needs implementation in both Core and GUI loader)

**Comment by peterNordin** [comment]
I will review further some other day, but this is a bigger change than just "plotting" related, so we need to think it through carefully.

File: componentLibraries/defaultLibrary/Signal/Control/SignalPID2.hpp [hunk 4]
```diff
@@ -130,7 +130,7 @@ class SignalPID2 : public ComponentSignal
         mI = mI + mTimestep/mTt * (u-v);
 
         // Write output and remember error
-        (*mpOut) = v;
+        (*mpOut) = u;
```
[the remark this answers is not in the document]
Comment (MarkusBagge) [reply componentLibraries/defaultLibrary/Signal/Control/SignalPID2.hpp]: Since the parameters was named Umin and Umax I assumed they refered not only to the I-part but also to the output signal u. I would guess it's at least a common (most common?) use case to limit the output also. Third alternative might be to have separate limits for integrator and output. Better documentation in any way would sure help.

File: HopsanCore/src/Quantities.cpp (hunk 2 above)
Comment (MarkusBagge) [reply HopsanCore/src/Quantities.cpp]: Haha, yes I made the same Googling. The reason is cosmetic; units mOhm, kOhm, MOhm etc becomes easier to read than mohm, kohm and Mohm. Besides, I think it's not recommended with short prefix at all. Instead milliohm and kiloohm should be used. Personlly, I was just fine with mOhm, kOhm etc. Easy to read and understood by most.

File: HopsanCore/src/Quantities.cpp (hunk 3 above)
Comment (MarkusBagge) [reply HopsanCore/src/Quantities.cpp]: Hz was not used as base unit before. So I guess that can't be mixed up (hopefully). However, the other way round you are right I guess. If a variable was added - e.g. with addInputVariable() - with quantity "Frequency" instead of unit "Hz" or "1/s", then it will be troublesome. Because Frequency was rad/s before and now is Hz in the commit. 
Just checked; there are actually a lot of components using "Frequency", for instance Hydraulic33Valve.hpp. I.e. these components uses rad/s and not Hz I guess. 
So yes, you are right, we have a problem there. Solution? 
- Two different quantites Frequency=rad/s and FrequencyHz=Hz?
- Update components that used Frequency to use AngularVelocity instead? Both were rad/s before. Then let Frequency be Hz.
- A ton of other alternatives? 

Hmm... It seams like components have a bit too much freedom to select quantities and units for the moment.

File: componentLibraries/defaultLibrary/Signal/Control/SignalPID2.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (robbr48) [reply componentLibraries/defaultLibrary/Signal/Control/SignalPID2.hpp]: The constants Umin and Umax should at least have a description. We could also make it an option whether or not the output signal should be limited or only the I-part.

File: componentLibraries/defaultLibrary/Electric/ElectricACmachine.hpp (hunk 1 above)
Comment (robbr48) [reply componentLibraries/defaultLibrary/Electric/ElectricACmachine.hpp]: That is correct, but I believe Peter meant to write "Mathematica" :) We have a Modelica generator as well, but I don't think any component in the default library is currently generated from there.

Anyway, both "Frequency" and "Hz" are better than "1/s", since it is unclear if "1/s" is a frequency or an angular frequency.

File: HopsanCore/src/Quantities.cpp (hunk 3 above)
Comment (robbr48) [reply HopsanCore/src/Quantities.cpp]: > Two different quantites Frequency=rad/s and FrequencyHz=Hz?

In that case it should be called AngularFrequency and Frequency I think. But that would also be annoying, since you may want to convert unit from rad/s to Hz on the same signal.

File: componentLibraries/defaultLibrary/Hydraulic/Special/HydraulicPiston.hpp [hunk 5]
```diff
@@ -162,7 +162,7 @@ class HydraulicPiston : public ComponentC
 0.001,&mpV02);
             addInputVariable("SL", "Stroke", "m", 0.5,&mpSL);
             addInputVariable("betae", "Bulk modulus", "Pa", 1.7e7,&mpbetae);
-            addInputVariable("alpha", "Damp. factor", "1/s ", 0.,&mpalpha);
+            addInputVariable("alpha", "Damp. factor", "Hz", 0.,&mpalpha);
```
[the remark this answers is not in the document]
Comment (peterNordin) [reply componentLibraries/defaultLibrary/Hydraulic/Special/HydraulicPiston.hpp]: Initially the addInputVariable() function took unit as string. When I added quantities years later I did not want to break the API compatibility. So I just used the existing unit parameter and allowed QuantityOrUnit. So if you have a strange unit that there is no quantity for it will display that. It would be good if all components were updated to use Quantity, except when they use a non-base unit. The intention was for all simulation core component variables to use base units (SI units) whenever possible.

And yes using "strongly typed" quantities would be better, but it would require rewriting all components. But sure such a system could be designed, and the old API kept for backwards compatibility. Right now it would not add much since all existing quantitis are unique strings. Also the "auto update old models" problem remains

[the remark this answers is not in the document]
Comment (robbr48) [reply componentLibraries/defaultLibrary/Hydraulic/Special/HydraulicPiston.hpp]: I don't like the idea of using SI units for all parameters. 

First, there will be obvious problems for the frequency quantity, where both rad/s and Hz are actually SI units. The only solution would be to differ between "frequency" and "angular frequency", but you may still want to convert between the two units. 

Second, in many cases the "standard" parameters are not in SI units. In aircraft design, for example, they often use different units for distinguishing different parameters from each other, such as length, width, altitude etc. 

In conclusion, I like the idea of specifying a quantity, but it should be possible to change the default unit as well, for example by a colon syntax like "Position:inch". It should be the responsibility of the creator of the component to convert to SI units in the component code.

**Comment by peterNordin** [comment]
I think that this pull request should be split into three new ones. (But we keep this since there is a lot of discussion). 

Cretate new branches from "hopsan/master" and then 'git cherrypick' the relevant commits from this branch.

# One for the ohm additions
# Second for resolve quantity and frequency changes. 
# Third (if you want) for improving pid2, otherwise I can do it when I "have time"

File: componentLibraries/defaultLibrary/Hydraulic/Special/HydraulicPiston.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (peterNordin) [reply componentLibraries/defaultLibrary/Hydraulic/Special/HydraulicPiston.hpp]: The intention was that you should allways specify Quantity instead of unit. Then if you choose Frequency, only rad/s and *Hz are shown, it does not make sense to show, "rpm" and such units. Likewise for Ang.Vel, showing *Hz does not make sense. I have never seen rotation speed in that unit. rps (same as Hz) or rpm or rad/s are valid. And it is still possible to add custom units if you want *Hz.

Regarding not using SI units, I partially aggre, but mostly not :) I will open a new issue and copy some comments from this thread and we can continue that discussion there. I think that some (re)design is needed.

[the remark this answers is not in the document]
Comment (peterNordin) [reply componentLibraries/defaultLibrary/Hydraulic/Special/HydraulicPiston.hpp]: The problem comes when only unit is given (as it was from the beginning), then the Quantity can not be deduced automatically. But I think that was already clear.

**Comment by MarkusBagge** [comment]
New pull request #1797 created for the SignalPID2. Just a proposal.

**Comment by MarkusBagge** [comment]
New pull request #1798 created for Ohm.

File: componentLibraries/defaultLibrary/Hydraulic/Special/HydraulicPiston.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (MarkusBagge) [reply componentLibraries/defaultLibrary/Hydraulic/Special/HydraulicPiston.hpp]: Yes, because same base unit is used for different quantities. But is it vice to allow that?

Assuming we want (I do) both rad/s and Hz as (base) units when writing components, there should be two different quantities. Using rad/s for AngularVelocity and Hz for Frequency feels most intuitive to me. But perhaps we are forced to map Hz to something else like FrequencyHz (or whatever) for compability reasons, see below.

Even if Hz is not a commonly used unit for AngularVelocity I see no real problem to add Hz as custom unit to AngularFrequency. The unit selection dropdown lists in the GUI becomes a bit longer, that's all. This allow users to select an appropriate custom unit also for old models where rad/s may refer to both angular velocity and frequency.

> Likewise for Ang.Vel, showing *Hz does not make sense. I have never seen rotation speed in that unit. rps (same as Hz) or rpm or rad/s are valid.  

Well, not common but sure possible. Actually, I've experienced Hz as rotational speed too. In my opinion, there is little reason to not provide conversion between Hz and rad/s.

Then about redefining the base unit for quantity Frequency from rad/s to Hz. Problem occurs when quantity Frequency is specified in old components instead of rad/s. We may have a converter that rewrites Frequency to rad/s if hopsanguiversion<2.xx? I guess there is an caveat though...

Components in external libraries, i.e. that does not come with a new release of Hopsan, may still assume that Frequency means rad/s. Is there a way to convert these and do they (or the library) have any version info to determine if it's needed? 
This is why we may want to map base unit Hz to a new quantity like FrequencyHz. That shouldn't break any backward compability. Then Frequency (rad/s) could be made an alias for AngularVelocity too.

Finally, a reflection; Not sure what actually motivates the explicit use of quantities...
- When writing components; reconsider using units instead of quantities?
-- Quantities can be misinterpreted, units can't. 
-- A component author must know the unit anyhow, hence a reverse unit lookup must be manually performed by the author if only quantities are accepted as input. For instance; would everyone assume Frequency is rad/s?
- Suggestion; let all units that are "compatible" i.e. can be transformed by new_unit=old_unit*k+m have the same quantity. I.e. there should actually be no need for both frequency and angular velocity.
- In theory - but maybe not motivated - one could even get rid of the explicit use of base units all together. Because if all units within a quantity are "compatible" then conversion can always be done between them. It would allow us to write components and specify both kHz, rad/s, deg/s etc and Hopsan can handle it. Not saying it's a good idea, just possible. It would be more flexible though and require no extra knowledge of Hopsan-quantites, for good and bad.

Sorry for my tedious thoughts... Feel free to disregard all cause they tend to be less than perfect too often also :)

File: HopsanCore/src/Quantities.cpp (hunk 3 above)
Comment (robbr48) [reply HopsanCore/src/Quantities.cpp]: Isn't rad/s commonly used in control theory, while Hz is more common in signal processing? Then one or the other of these disciplines will be annoyed if we force everyone to use the same default unit.

File: componentLibraries/defaultLibrary/Hydraulic/Special/HydraulicPiston.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (MarkusBagge) [reply componentLibraries/defaultLibrary/Hydraulic/Special/HydraulicPiston.hpp]: Well yes, I guess quantity was a poor designation for "compatible" units. By component modeller, do you mean the person who writes component code? The component code may specify any of "Hz" or "rad/s". Then what unit will be prompted/plotted just depends on what was chosen by the user in the Hopsan default/custom units dialogue. If you mix control and mechanical components in a model though, the same default would be used which can be, as mentioned, quirky. But still, you would be able to change to Hz or rad/s etc in the unit dropdown list.

Anyhow, agree that the more sophisticated approach with quantities may offers advantages in terms of usability. There is a nice list here b.t.w. https://en.wikipedia.org/wiki/List_of_physical_quantities. 

So if we stick to using quantities. Should full backwards compatibility be kept, including external libraries? Then I guess we can't redefine quantity Frequency to use Hz? Reasons mentioned in previous posts. Any reasonable workaround, e.g. inventing a new quantity "FrequencySomethingElse" and use for Hz.

File: HopsanCore/src/Quantities.cpp (hunk 3 above)
Comment (MarkusBagge) [reply HopsanCore/src/Quantities.cpp]: Hmm... You don't want to stand in between those camps I guess, so maybe we should avoid starting a war :-)

File: componentLibraries/defaultLibrary/Hydraulic/Special/HydraulicPiston.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (peterNordin) [reply componentLibraries/defaultLibrary/Hydraulic/Special/HydraulicPiston.hpp]: Some background. My initial implementation only had si units (baseunit) and scale units. But then I ran into various problems (do not remember exactly), among them,  this trouble with the frequency unit and mixup with angular velocity, and there were cases where converting between units in the two quantities would cause incorrect values (do not remember exactly how). There were also cases where sometimes rad/s and sometimes Hz were used. It was a mess.

At the time I was also doing some work on fmi, so I was reading the standard https://fmi-standard.org/docs/2.0.1-develop/#_definition_of_units_unitdefinitions 
And found what they were doing and decided to do something similar. 

They also mention the rad/s, Hz problem, and mention some other as well.
They also want to make it possible to check if a connection is valid (compatible units) and especially for non-standard units making sure is difficult. So “quantities" at least allow comparison  if two ports work with the same quantity and the same unit then they should be compatible.

This is not used in Hopsan (yet). But if such checking (quantity and unit) was added then maybe we could even allow non base (SI or SI derived) units.
Maybe on the form Frequency::Hz as Robert sugested, then we could not connect such a port to Frequency::rad/s.

This could also be used for default scaling of values in parameters (constants and start values).

But I think that there could still be problems with non SI base units. If you write a sctipt for instance, and have not checked the component you may be supprised to find that you should set a length in mm instead of the expected m.   

The wikipedia link is good, but we should also look at the nist guidelines I linked elsewhere and the fmi spec.

Ensuring backward compatibility if we change base uniy is difficult sin we can not change others external component libraries.

[the remark this answers is not in the document]
Comment (MarkusBagge) [reply componentLibraries/defaultLibrary/Hydraulic/Special/HydraulicPiston.hpp]: Hmm... confusing version info on the FMI website. FMI version 2 is dated 2014 while version 1 is dated 2017. In the downloads section, for version 2.01 there is a link to a GitHub repository. There, I went checking the .xsd schemas at https://github.com/modelica/fmi-standard/tree/master/schema to understand the structure but noticed the file is named fmi3Unit.xsd (version 3!?!) rather than fmi2Unit.xsd. Probably missing some insights here...

However, I notice that the FMI spec allows specifying radians "rad" as if it would have been a SI-base unit. This would be legal;
```xml
<unit name="AngularVelocity">  <!-- FMI spec use rad/s as example but any unique string is allowed as name -->
  <BaseUnit kg="0" m="0" s="-1" A="0" K="0" mol="0" cd="0" rad="1" factor="1" offset="0" />
  <DisplayUnit name="rad/s" factor="1" offset="0" /> 
  <DisplayUnit name="rpm" factor="0.1047" offset="0" />   <!-- factor=2*pi/60 -->
</unit>
<unit name="Frequency">
  <BaseUnit kg="0" m="0" s="-1" A="0" K="0" mol="0" cd="0" rad="0" factor="1" offset="0" />
  <DisplayUnit name="Hz" factor="1" offset="0" />
  <DisplayUnit name="kHz" factor="1e3" offset="0" />
</unit>
```
Same thing in "Hopsan-style";
```xml
<quantity name="AngularVelocity" baseunit="rad/s">
  <unit name="rpm" scale="2*pi/60"/>
</quantity>
<quantity name="Frequency" baseunit="Hz">
  <unit name="kHz" scale="1e-3"/>
</quantity>
```

Sure there are similarites here; 
- FMI unit name attribute string <-> Hopsan quantity name attribute string and 
- FMI base unit attributes <-> Hopsan base unit string


Anyhow... So first question is if backward compatibility for Frequency should be kept. Should it?
If no: 
- We reassign "Hz" to quantity Frequency and update components and models that "ships" with Hopsan. (Old models, external libraries etc that has specified quantity Frequency might fail.)

If yes: 
- We assign "Hz" to a new, never used before, quantity as a workaround. Or...
- We assign "Hz" to Frequency and make Hopsan convert old models/components (as far as possible) by swapping quantity Frequency for AngularFrequency. Unexpected units may be displayed in plots/menus but results will remain numerically correct (right so?) since base unit "rad/s" is keept by specifying AngularVelocity. Authors may choose to rewrite their stuff in case they really want the custom units of Frequency to become available in plots/menus. 

This is mostly about improving Hopsan user friendliness. It has been enormously improved last years but there are still some small improvements I think. I will try to help if I am able to. But for this particular matter, I leave it with a humble wish for future fine tuning. Feel free to deny this pull request until there is a decided way. Thanks!

[the remark this answers is not in the document]
Comment (peterNordin) [reply componentLibraries/defaultLibrary/Hydraulic/Special/HydraulicPiston.hpp]: I think that I have forgotten the original reason to why you want to have Hz as the base (or default) Frequency unit. Can you give an example ? Or was it due to the collission with AngularVelocity.  Lets move any further discussion to #1801, but you can reply to this question here. I will update #1801 with any new information.

[the remark this answers is not in the document]
Comment (robbr48) [reply componentLibraries/defaultLibrary/Hydraulic/Special/HydraulicPiston.hpp]: Hz is the base unit for frequency, while rad/s is the base unit for angular frequency/velocity. I guess the confusing part is that with angular velocity, you can convert from rpm to  Hz, but with frequency you cannot convert Hz to rpm (for example). I'm not sure if we should also separate "angular frequency" from "angular velocity"? It does not make sense to have a filter with a break frequency in rpm, but still rad/s is a better default unit than Hz in this case.

[the remark this answers is not in the document]
Comment (MarkusBagge) [reply componentLibraries/defaultLibrary/Hydraulic/Special/HydraulicPiston.hpp]: Sorry for the delay. Some busy days... 
Original reason 1; The GUI unit handling is inconsistent. Sometimes user can select custom units, sometimes not. 
In the attempts to find out why, it was noticed 
a) Components many times states units rather than quantities when defining variables. Example ElectricMotor.hpp;
```
addInputVariable("wc", "Friction speed", "rad/s", 1.,&mpwc);
```
b) Hopsan select custom units based on those defined in the hopsandefaults xml file. System must know the quantity to do the reverse lookup.
c) Such 1:1 reverse lookup unit->quantity is ambiguous since rad/s can be both Frequency and AngularVelocity (in both hopsandefaults and quantity.cpp). 

Conclusion; Separate Frequency from AngularVelocity so they don't use the same base unit! That creates confusion, Hopsan won't know which custom units to display in conversion menus for above reasons. And does it really make sense? 
Hz was a proposal because people would certainly associate Hz to frequency and rad/s to angular velocity.
The simple quick/simple/ugly proposed solution was to allow conversions between them all, let it make sense or not. Nothing would prevent an end user to select a sensible unit in a list also containing some less sensible ones. Early suggestion in pull request was to let Frequency just be an alias of AngularVelocity. But sure there are better ways.

Over and out :-)
